File: lib-database/src/main/java/com/ustadmobile/core/db/dao/SaleDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.core.db.UmLiveData;
import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmInsert;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmRepository;
import com.ustadmobile.lib.database.annotation.UmUpdate;
import com.ustadmobile.lib.db.entities.Sale;
import com.ustadmobile.lib.db.entities.SaleListDetail;
import com.ustadmobile.lib.db.sync.dao.SyncableDao;

import java.util.List;



@UmDao(updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN, insertPermissionCondition =
RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
public abstract class SaleDao implements SyncableDao<Sale, SaleDao> {

    public static final int ALL_SELECTED = 1;
    public static final int PREORDER_SELECTED = 2;
    public static final int PAYMENT_SELECTED = 3;

    public static final int SORT_ORDER_NAME_ASC =1;
    public static final int SORT_ORDER_NAME_DESC =2;
    public static final int SORT_ORDER_AMOUNT_ASC=3;
    public static final int SORT_ORDER_AMOUNT_DESC=4;
    public static final int SORT_ORDER_DATE_CREATED_DESC=5;
    public static final int SORT_ORDER_DATE_CREATED_ASC=6;


    //INSERT

    @UmInsert
    public abstract void insertAsync(Sale entity, UmCallback<Long> insertCallback);


    //FIND ALL ACTIVE

    public static final String ALL_SALES_QUERY = "SELECT * FROM Sale";

    @UmQuery(ALL_SALES_QUERY)
    public abstract List<Sale> findAllList();

    public static final String ALL_SALES_ACTIVE_QUERY = "SELECT * FROM Sale WHERE saleActive = 1";

    @UmQuery(ALL_SALES_ACTIVE_QUERY)
    public abstract UmLiveData<List<Sale>> findAllActiveLive();

    @UmQuery(ALL_SALES_ACTIVE_QUERY)
    public abstract List<Sale> findAllActiveList();

    @UmQuery(ALL_SALES_ACTIVE_QUERY)
    public abstract void findAllActiveAsync(UmCallback<List<Sale>> allActiveSalesCallback);

    @UmQuery(ALL_SALES_ACTIVE_QUERY)
    public abstract UmProvider<Sale> findAllActiveProvider();

    @UmQuery(ALL_SALES_ACTIVE_QUERY +
            " AND salePreOrder = 1 ")
    public abstract UmLiveData<List<Sale>> findAllActivePreorderSalesLive();

    @UmQuery(ALL_SALES_ACTIVE_QUERY +
            " AND salePaymentDone = 0")
    public abstract UmLiveData<List<Sale>> findAllActivePaymentDueSalesLive();

    @UmQuery("SELECT * FROM Sale WHERE saleTitle = :saleTitle AND saleActive = 1")
    public abstract void findAllSaleWithTitleAsync(String saleTitle, UmCallback<List<Sale>> resultCallback);

    @UmQuery("SELECT * FROM Sale WHERE saleTitle = :saleTitle AND saleActive = 1")
    public abstract List<Sale> findAllSaleWithTitle(String saleTitle);


    /*
    String saleTitle;
    String locationName;
    long saleDueDate;
    float saleAmount;
    String saleCurrency;
    long saleCreationDate;
    int saleItemCount;
     */
//    public static final String ALL_SALE_LIST =
//    " SELECT s.*, Location.title AS locationName, " +
//    " COALESCE( (SELECT SUM(SaleItem.saleItemPricePerPiece * SaleItem.saleItemQuantity) - " +
//            "SUM(Sale.saleDiscount)  FROM Sale LEFT JOIN SaleItem on SaleItem.saleItemSaleUid = " +
//            "Sale.saleUid WHERE Sale.saleUid = s.saleUid) ,0) AS saleAmount, " +
//    " 'Afs' AS saleCurrency,  " +
//    " (SELECT count(*) FROM SaleItem WHERE SaleItem.saleItemSaleUid = s.saleUid) AS saleItemCount " +
//    "FROM Sale s " +
//    " LEFT JOIN Location ON Location.locationUid = s.saleLocationUid WHERE s.saleActive = 1";
//
    public static final String ALL_SALE_LIST =
            " SELECT sl.*, " +
            " (SELECT SaleItem.saleItemQuantity " +
            " FROM Sale stg " +
            " LEFT JOIN SaleItem ON SaleItem.saleItemSaleUid = stg.saleUid " +
            " WHERE stg.saleUid = sl.saleUid " +
            " ORDER BY stg.saleCreationDate ASC LIMIT 1 " +
            " )  " +
            " || 'x ' || " +
            " (SELECT SaleProduct.saleProductName " +
            " FROM SaleItem sitg " +
            " LEFT JOIN SaleProduct ON SaleProduct.saleProductUid = sitg.saleItemProductUid " +
            " WHERE sitg.saleItemSaleUid = sl.saleUid " +
            " ORDER BY sitg.saleItemCreationDate ASC LIMIT 1) " +
            " || " +
            " (select " +
            "  (case  " +
            "   when  " +
            "   (SELECT count(*) from SaleItem sid where sid.saleItemSaleUid = sl.saleUid) > 1 " +
            "   then '...'  " +
            "   else '' " +
            "  end) " +
            " from sale) " +
            " AS saleTitleGen, " +
            " Location.title AS locationName, " +
            " COALESCE( (SELECT SUM(SaleItem.saleItemPricePerPiece * SaleItem.saleItemQuantity) - " +
            "            SUM(Sale.saleDiscount)  FROM Sale LEFT JOIN SaleItem on SaleItem.saleItemSaleUid = " +
            "            Sale.saleUid WHERE Sale.saleUid = sl.saleUid) ,0 " +
            " ) AS saleAmount, " +
            " (COALESCE( (SELECT SUM(SaleItem.saleItemPricePerPiece * SaleItem.saleItemQuantity) - " +
            "            SUM(Sale.saleDiscount)  FROM Sale LEFT JOIN SaleItem on SaleItem.saleItemSaleUid = " +
            "            Sale.saleUid WHERE Sale.saleUid = sl.saleUid) ,0 " +
            " ) - COALESCE((SELECT SUM(SalePayment.salePaymentPaidAmount) FROM SalePayment " +
            " WHERE SalePayment.salePaymentSaleUid = sl.saleUid " +
            " AND SalePayment.salePaymentDone = 1 AND SalePayment.salePaymentActive = 1) ,0)) AS saleAmountDue, " +
            " 'Afs' AS saleCurrency,  " +
            " coalesce(" +
            "    ( " +
            "    SELECT SaleItem.saleItemDueDate FROM SaleItem LEFT JOIN Sale on Sale.saleUid = " +
            "       SaleItem.saleItemSaleUid WHERE SaleItem.saleItemSaleUid = sl.saleUid  " +
            "      AND Sale.salePreOrder = 1 AND Sale.saleActive = 1 AND SaleItem.saleItemPreOrder = 1 " +
            "      AND SaleItem.saleItemSold = 0 ORDER BY SaleItem.saleItemDueDate ASC LIMIT 1 " +
            "    ) ,0) AS earliestDueDate, "  +
            " (SELECT count(*) FROM SaleItem WHERE SaleItem.saleItemSaleUid = sl.saleUid) AS saleItemCount," +
            " COALESCE((SELECT SUM(SalePayment.salePaymentPaidAmount) FROM SalePayment  " +
            "  WHERE SalePayment.salePaymentSaleUid = sl.saleUid " +
            "  AND SalePayment.salePaymentDone = 1 AND SalePayment.salePaymentActive = 1) ,0) " +
            "  AS saleAmountPaid " +
            " FROM Sale sl " +
            " LEFT JOIN Location ON Location.locationUid = sl.saleLocationUid WHERE sl.saleActive = 1  ";



    @UmQuery(ALL_SALE_LIST)
    public abstract UmLiveData<List<SaleListDetail>> findAllActiveAsSaleListDetailLive();

    @UmQuery(ALL_SALE_LIST)
    public abstract List<SaleListDetail> findAllActiveAsSaleListDetailList();

    @UmQuery(ALL_SALE_LIST)
    public abstract void findAllActiveAsSaleListDetailAsync(UmCallback<List<SaleListDetail>> allActiveSalesCallback);

    @UmQuery(ALL_SALE_LIST + " AND salePreOrder = 1")
    public abstract UmLiveData<List<SaleListDetail>> findAllActiveSaleListDetailPreOrdersLive();

    @UmQuery(ALL_SALE_LIST + " AND salePaymentDone = 1")
    public abstract UmLiveData<List<SaleListDetail>> findAllActiveSaleListDetailPaymentDueLive();

    //filter and sort

    public static final String FILTER_PREORDER = " AND salePreOrder = 1";
    public static final String FILTER_PAYMENT_DUE = " AND saleAmountPaid < saleAmount ";

    @UmQuery(ALL_SALE_LIST)
    public abstract UmProvider<SaleListDetail> findAllActiveAsSaleListDetailProvider();

    @UmQuery(ALL_SALE_LIST + FILTER_PREORDER)
    public abstract UmProvider<SaleListDetail> findAllActiveSaleListDetailPreOrdersProvider();


    //Payments due shows the payment amount pending vs the total amount of the sale.
    @UmQuery(ALL_SALE_LIST + FILTER_PAYMENT_DUE)
    public abstract UmProvider<SaleListDetail> findAllActiveSaleListDetailPaymentDueProvider();




    public static final String SORT_NAME_ASC = " ORDER BY sl.saleTitle ASC ";
    public static final String SORT_NAME_DEC = " ORDER BY sl.saleTitle DESC ";
    public static final String SORT_TOTAL_AMOUNT_DESC = " ORDER BY saleAmount DESC ";
    public static final String SORT_TOTAL_AMOUNT_ASC = " ORDER BY saleAmount ASC ";
    public static final String SORT_ORDER_DATE_DESC = " ORDER BY sl.saleCreationDate DESC ";
    public static final String SORT_ORDER_DATE_ASC = " ORDER BY sl.saleCreationDate ASC ";

    @UmQuery(ALL_SALE_LIST +  SORT_NAME_ASC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterAllSortNameAscProvider();

    @UmQuery(ALL_SALE_LIST + SORT_NAME_DEC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterAllSortNameDescProvider();

    @UmQuery(ALL_SALE_LIST + SORT_TOTAL_AMOUNT_DESC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterAllSortTotalAscProvider();

    @UmQuery(ALL_SALE_LIST + SORT_TOTAL_AMOUNT_ASC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterAllSortTotalDescProvider();

    @UmQuery(ALL_SALE_LIST + SORT_ORDER_DATE_DESC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterAllSortDateAscProvider();

    @UmQuery(ALL_SALE_LIST + SORT_ORDER_DATE_ASC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterAllSortDateDescProvider();

    @UmQuery(ALL_SALE_LIST + FILTER_PREORDER + SORT_NAME_ASC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterPreOrderSortNameAscProvider();

    @UmQuery(ALL_SALE_LIST + FILTER_PREORDER + SORT_NAME_DEC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterPreOrderSortNameDescProvider();

    @UmQuery(ALL_SALE_LIST + FILTER_PREORDER + SORT_TOTAL_AMOUNT_DESC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterPreOrderSortTotalAscProvider();

    @UmQuery(ALL_SALE_LIST + FILTER_PREORDER + SORT_TOTAL_AMOUNT_ASC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterPreOrderSortTotalDescProvider();

    @UmQuery(ALL_SALE_LIST + FILTER_PREORDER + SORT_ORDER_DATE_DESC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterPreOrderSortDateAscProvider();

    @UmQuery(ALL_SALE_LIST + FILTER_PREORDER + SORT_ORDER_DATE_ASC)
    public abstract UmProvider<SaleListDetail> findAllSaleFilterPreOrderSortDateDescProvider();


    public UmProvider<SaleListDetail> filterAndSortSale(int filter, int sort){

        switch (filter){
            case ALL_SELECTED:
                switch (sort){
                    case SORT_ORDER_NAME_ASC:
                        return findAllSaleFilterAllSortNameAscProvider();
                    case SORT_ORDER_NAME_DESC:
                        return findAllSaleFilterAllSortNameDescProvider();
                    case SORT_ORDER_AMOUNT_ASC:
                        return findAllSaleFilterAllSortTotalAscProvider();
                    case SORT_ORDER_AMOUNT_DESC:
                        return findAllSaleFilterAllSortTotalDescProvider();
                    case SORT_ORDER_DATE_CREATED_DESC:
                        return findAllSaleFilterAllSortDateAscProvider();
                    case SORT_ORDER_DATE_CREATED_ASC:
                        return findAllSaleFilterAllSortDateDescProvider();
                }
                break;
            case PREORDER_SELECTED:
                switch (sort){
                    case SORT_ORDER_NAME_ASC:
                        return findAllSaleFilterPreOrderSortNameAscProvider();
                    case SORT_ORDER_NAME_DESC:
                        return findAllSaleFilterPreOrderSortNameDescProvider();
                    case SORT_ORDER_AMOUNT_ASC:
                        return findAllSaleFilterPreOrderSortTotalAscProvider();
                    case SORT_ORDER_AMOUNT_DESC:
                        return findAllSaleFilterPreOrderSortTotalDescProvider();
                    case SORT_ORDER_DATE_CREATED_DESC:
                        return findAllSaleFilterPreOrderSortDateAscProvider();
                    case SORT_ORDER_DATE_CREATED_ASC:
                        return findAllSaleFilterPreOrderSortDateDescProvider();
                }
                break;
            case PAYMENT_SELECTED:
                break;
        }
        return findAllActiveAsSaleListDetailProvider();
    }

    //LOOK UP

    public static final String FIND_BY_UID_QUERY = "SELECT * FROM Sale WHERE saleUid = :saleUid";

    @UmQuery(FIND_BY_UID_QUERY)
    public abstract Sale findByUid(long saleUid);

    @UmQuery(FIND_BY_UID_QUERY)
    public abstract void findByUidAsync(long saleUid, UmCallback<Sale> findByUidCallback);

    @UmQuery(FIND_BY_UID_QUERY)
    public abstract UmLiveData<Sale> findByUidLive(long saleUid);

    //INACTIVATE:

    public static final String INACTIVATE_SALE_QUERY =
            "UPDATE Sale SET saleActive = 0 WHERE saleUid = :saleUid";
    @UmQuery(INACTIVATE_SALE_QUERY)
    public abstract void inactivateEntity(long saleUid);

    @UmQuery(INACTIVATE_SALE_QUERY)
    public abstract void inactivateEntityAsync(long saleUid, UmCallback<Integer> inactivateSaleCallback);


    //UPDATE:

    @UmUpdate
    public abstract void updateAsync(Sale entity, UmCallback<Integer> updateCallback);


    //Get overdue sale count
    @UmQuery("select count(*) from sale where Sale.saleDueDate < :today and Sale.saleDueDate > 0 AND Sale.saleActive = 1")
    public abstract void getOverDueSaleCountAsync(long today, UmCallback<Integer> resultCallback);

    @UmQuery("select count(*) from sale where salePreOrder = 1 AND saleActive = 1")
    public abstract UmProvider<Integer> getPreOrderSaleCountProvider();

    @UmQuery("select count(*) from sale where salePreOrder = 1 AND saleActive = 1")
    public abstract UmLiveData<Integer> getPreOrderSaleCountLive();
}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/SaleDao.kt
package com.ustadmobile.core.db.dao

import androidx.paging.DataSource
import androidx.room.Dao
import androidx.room.Query
import androidx.room.Update
import com.ustadmobile.lib.db.entities.*
import com.ustadmobile.door.annotation.Repository

@Repository
@Dao
abstract class SaleDao : BaseDao<Sale> {

    @Update
    abstract suspend fun updateAsync(entity: Sale): Int

    //Used in SaleEditPresenter to find customer and location to update on sale edit
    @Query(FIND_SALE_WITH_CUSTOMER_AND_LOCATION)
    abstract suspend fun findWithCustomerAndLocationByUidAsync(uid: Long): SaleWithCustomerAndLocation?

    //Query to find all sales in SaleList
    @Query(FIND_ALL_SALES)
    abstract fun findAllSales(
            leUid: Long, filter: Int, searchText: String): DataSource.Factory<Int,SaleListDetail>

    //People tab at bottom navigation - Show all people with Sale info (if any, eg: total sales)
    @Query(FIND_ALL_PEOPLE_WITH_SALEINFO)
    abstract fun findAllPersonWithSaleInfo(leUid: Long, filter: Int,
                                           searchText: String) : DataSource.Factory<Int, PersonWithSaleInfo>

    companion object {

        const val FILTER_ALL = 0
        const val FILTER_LE_ONLY = 1
        const val FILTER_WE_ONLY = 2
        const val FILTER_CUSTOMER_ONLY = 3

        const val FILTER_ALL_SALES = 0
        const val FILTER_PAYMENTS_DUE_SALES = 1


        const val SELECT_ACCOUNT_IS_ADMIN = "(SELECT admin FROM Person WHERE personUid = :accountPersonUid)"

        const val FIND_SALE_WITH_CUSTOMER_AND_LOCATION = """
            SELECT Sale.*, Person.*, Location.*  FROM Sale 
            LEFT JOIN Person ON Person.personUid = Sale.saleCustomerUid 
            LEFT JOIN Location ON Location.locationUid = Sale.saleLocationUid
            WHERE Sale.saleUid = :uid AND CAST(Sale.saleActive AS INTEGER) = 1 
        """

        const val FIND_ALL_PEOPLE_WITH_SALEINFO = """
            SELECT 
                coalesce(SUM((SaleItem.saleItemPricePerPiece) * SaleItem.saleItemQuantity) - 
                    (SUM(SaleItem.saleItemDiscount) + SUM(Sale.saleDiscount)), 0) AS totalSale,
                '' AS topProducts, 
                0 as personPictureUid ,
                Person.* 
            FROM Person 
                LEFT JOIN Person AS LE ON LE.personUid = :leUid
                LEFT JOIN InventoryItem ON 
                    InventoryItem.InventoryItemWeUid = Person.personUid 
                    AND InventoryItem.InventoryItemLeUid = LE.personUid AND CAST(InventoryItem.inventoryItemActive AS INTEGER) = 1
                LEFT JOIN InventoryTransaction ON 
                    InventoryTransaction.InventoryTransactionInventoryItemUid = InventoryItem.InventoryItemUid 
                    AND InventoryTransaction.inventoryTransactionFromLeUid = LE.personUid AND CAST(InventoryTransaction.inventoryTransactionActive AS INTEGER) = 1
                LEFT JOIN Sale ON Sale.salePersonUid = Person.personUid
                    AND CAST(Sale.saleActive AS INTEGER) = 1
                 LEFT JOIN SaleItem ON    SaleItem.saleItemSaleUid = Sale.saleUid 
                    AND CAST(SaleItem.saleItemActive AS INTEGER) = 1               
                LEFT JOIN Product ON Product.productUid = SaleItem.saleItemProductUid 
                    AND CAST(Product.productActive AS INTEGER) = 1
            WHERE CAST(Person.active AS INTEGER) = 1  
            AND (CAST(Person.admin AS INTEGER) = 0 OR CAST(LE.admin AS INTEGER) = 1 )
            AND (Person.personCreatedBy = LE.personUid OR CAST(LE.admin AS INTEGER) = 1)
            AND CASE :filter WHEN $FILTER_ALL THEN Person.personGoldoziType > -1 
                ELSE CAST(Person.admin AS INTEGER) = 0 END
            AND CASE :filter WHEN $FILTER_LE_ONLY THEN Person.personGoldoziType = 1
                ELSE Person.personGoldoziType > -1 END
            AND CASE :filter WHEN $FILTER_WE_ONLY THEN Person.personGoldoziType = 2
                ELSE Person.personGoldoziType > -1 END 
            AND CASE :filter WHEN $FILTER_CUSTOMER_ONLY THEN Person.personGoldoziType = 0  
                ELSE Person.personGoldoziType > -1 END
            AND Person.firstNames ||' '|| Person.lastName LIKE :searchText
            GROUP BY Person.personUid
        """

        const val FIND_ALL_SALES = """
            SELECT sl.*, Customer.*, SalePerson.firstNames||' '||SalePerson.lastName as saleCreator,
                (SELECT SaleItem.saleItemQuantity 
                  FROM Sale stg 
                  LEFT JOIN SaleItem ON SaleItem.saleItemSaleUid = stg.saleUid 
                        AND CAST(SaleItem.saleItemActive AS INTEGER) = 1  
                  WHERE stg.saleUid = sl.saleUid AND CAST(SaleItem.saleItemActive AS INTEGER) = 1 
                  ORDER BY stg.saleCreationDate ASC LIMIT 1 
                  )  
                  || 'x ' || 
                  (SELECT case when Product.productName != '' 
                        then Product.productName 
                        else case when Product.productNameDari != '' 
                            then Product.productNameDari 
                            else case when Product.productNamePashto != '' 
                                then Product.productNamePashto 
                                else '' 
                                end  
                            end 
                        end 
                  FROM SaleItem sitg 
                  LEFT JOIN Product ON Product.productUid = sitg.saleItemProductUid 
                  WHERE sitg.saleItemSaleUid = sl.saleUid AND CAST(sitg.saleItemActive AS INTEGER) = 1  
                  ORDER BY sitg.saleItemCreationDate ASC LIMIT 1) 
                  || 
                  (select 
                      (case  
                      when  
                      (SELECT count(*) from SaleItem sid where sid.saleItemSaleUid = sl.saleUid 
                                and CAST(sid.saleItemActive AS INTEGER) = 1 ) > 1 
                      then '...'  
                      else '' 
                  end) 
                  from sale) 
                AS saleTitleGen, 
                
                (SELECT SaleItem.saleItemQuantity 
                  FROM Sale stg 
                  LEFT JOIN SaleItem ON SaleItem.saleItemSaleUid = stg.saleUid 
                    AND CAST(SaleItem.saleItemActive AS INTEGER) = 1  
                  WHERE stg.saleUid = sl.saleUid AND CAST(SaleItem.saleItemActive AS INTEGER) = 1 
                  ORDER BY stg.saleCreationDate ASC LIMIT 1 
                  )  
                  || 'x ' || 
                  (SELECT CASE WHEN Product.productNameDari IS NOT NULL AND Product.productNameDari != '' 
                    THEN Product.productNameDari ELSE Product.productName END 
                  FROM SaleItem sitg 
                  LEFT JOIN Product ON Product.productUid = sitg.saleItemProductUid 
                  WHERE sitg.saleItemSaleUid = sl.saleUid AND CAST(sitg.saleItemActive AS INTEGER) = 1  
                  ORDER BY sitg.saleItemCreationDate ASC LIMIT 1) 
                  || 
                  (select 
                      (case  
                      when  
                      (SELECT count(*) from SaleItem sid where sid.saleItemSaleUid = sl.saleUid 
                        and CAST(sid.saleItemActive AS INTEGER) = 1 ) > 1 
                      then '...'  
                      else '' 
                  end) 
                  from sale) 
                AS saleTitleGenDari, 
                
                (SELECT SaleItem.saleItemQuantity 
                  FROM Sale stg 
                  LEFT JOIN SaleItem ON SaleItem.saleItemSaleUid = stg.saleUid 
                    AND CAST(SaleItem.saleItemActive AS INTEGER) = 1  
                  WHERE stg.saleUid = sl.saleUid AND CAST(SaleItem.saleItemActive AS INTEGER) = 1 
                  ORDER BY stg.saleCreationDate ASC LIMIT 1 
                  )  
                  || 'x ' || 
                  (SELECT 
                    CASE WHEN Product.productNamePashto IS NOT NULL AND Product.productNamePashto != '' 
                    THEN Product.productNamePashto ELSE Product.productName END 
                  FROM SaleItem sitg 
                  LEFT JOIN Product ON Product.productUid = sitg.saleItemProductUid 
                  WHERE sitg.saleItemSaleUid = sl.saleUid AND CAST(sitg.saleItemActive AS INTEGER) = 1  
                  ORDER BY sitg.saleItemCreationDate ASC LIMIT 1) 
                  || 
                  (select 
                      (case  
                      when  
                      (SELECT count(*) from SaleItem sid where sid.saleItemSaleUid = sl.saleUid 
                        and CAST(sid.saleItemActive AS INTEGER) = 1 ) > 1 
                      then '...'  
                      else '' 
                  end) 
                  from sale) 
                AS saleTitleGenPashto, 
                
                '' AS productNames, 
                  
                '' AS productNamesDari,
                   
                '' AS locationName, 
                
                COALESCE( (SELECT SUM(SaleItem.saleItemPricePerPiece * SaleItem.saleItemQuantity) - 
                                SUM(Sale.saleDiscount)  
                            FROM Sale LEFT JOIN SaleItem on SaleItem.saleItemSaleUid = 
                                Sale.saleUid AND CAST(SaleItem.saleItemActive AS INTEGER) = 1  
                            WHERE Sale.saleUid = sl.saleUid) ,
                        0 ) AS saleAmount, 
                
                (0) AS saleAmountDue, 
                
                'Afs' AS saleCurrency,  
                
                coalesce(
                    (   
                        SELECT SaleItem.saleItemDueDate FROM SaleItem LEFT JOIN Sale on Sale.saleUid = 
                        SaleItem.saleItemSaleUid WHERE SaleItem.saleItemSaleUid = sl.saleUid  
                        AND CAST(Sale.saleActive AS INTEGER) = 1  
                        AND CAST(SaleItem.saleItemPreOrder AS INTEGER) = 1 
                        ORDER BY SaleItem.saleItemDueDate ASC LIMIT 1 
                    ) 
                , 0) AS earliestDueDate,
                     
                (SELECT count(*) FROM SaleItem WHERE SaleItem.saleItemSaleUid = sl.saleUid) AS saleItemCount,
                    COALESCE((SELECT SUM(SalePayment.salePaymentPaidAmount) FROM SalePayment  
                    WHERE SalePayment.salePaymentSaleUid = sl.saleUid 
                    AND CAST(SalePayment.salePaymentDone AS INTEGER) = 1 
                    AND CAST(SalePayment.salePaymentActive AS INTEGER) = 1 ) ,0) 
                AS saleAmountPaid,
                 
                (select (case  when  
                    (SELECT count(*) from SaleItem sip where sip.saleItemSaleUid = sl.saleUid 
                    AND CAST(sip.saleItemPreOrder AS INTEGER) = 1 ) > 0  then 1  else 0 end)  
                from Sale limit 1) AS saleItemPreOrder
                 
                FROM Sale sl 
                    LEFT JOIN Person AS Customer ON Customer.personUid = sl.saleCustomerUid
                    LEFT JOIN SaleItem ON SaleItem.saleItemSaleUid = sl.saleUid 
                        AND CAST(SaleItem.saleItemActive AS INTEGER) = 1
                    LEFT JOIN Person as LE ON LE.personUid = :leUid
                    LEFT JOIN Person as SalePerson ON SalePerson.personUid = sl.salePersonUid 
                
                WHERE CAST(sl.saleActive AS INTEGER) = 1
                
                AND (
					CAST(LE.admin AS INTEGER) = 1 OR 
					sl.salePersonUid = LE.personUid
					)

                AND             
                 (SELECT SaleItem.saleItemQuantity 
                  FROM Sale stg 
                  LEFT JOIN SaleItem ON SaleItem.saleItemSaleUid = stg.saleUid 
                    AND CAST(SaleItem.saleItemActive AS INTEGER) = 1  
                  WHERE stg.saleUid = sl.saleUid AND CAST(SaleItem.saleItemActive AS INTEGER) = 1 
                  ORDER BY stg.saleCreationDate ASC LIMIT 1 
                  )  
                  || 'x ' || 
                  (SELECT case when Product.productName != '' 
                        then Product.productName 
                        else case when Product.productNameDari != '' 
                            then Product.productNameDari 
                            else case when Product.productNamePashto != '' 
                                then Product.productNamePashto 
                                else '' 
                                end  
                            end 
                        end  	
                  FROM SaleItem sitg 
                  LEFT JOIN Product ON Product.productUid = sitg.saleItemProductUid 
                  WHERE sitg.saleItemSaleUid = sl.saleUid AND CAST(sitg.saleItemActive AS INTEGER) = 1  
                  ORDER BY sitg.saleItemCreationDate ASC LIMIT 1) 
                  || 
                  (select 
                      (case  
                      when  
                      (SELECT count(*) from SaleItem sid where sid.saleItemSaleUid = sl.saleUid 
                                and CAST(sid.saleItemActive AS INTEGER) = 1 limit 1) > 1 
                      then '...'  
                      else '' 
                  end) 
                  from sale limit 1) LIKE :searchText
                
                AND CASE :filter WHEN $FILTER_PAYMENTS_DUE_SALES THEN 
                    (COALESCE( (SELECT SUM(SaleItem.saleItemPricePerPiece * SaleItem.saleItemQuantity) - 
                           SUM(Sale.saleDiscount)  FROM Sale LEFT JOIN SaleItem on SaleItem.saleItemSaleUid = 
                           Sale.saleUid AND CAST(SaleItem.saleItemActive AS INTEGER) = 1  
                            WHERE Sale.saleUid = sl.saleUid) ,0 
                          ) - COALESCE((SELECT SUM(SalePayment.salePaymentPaidAmount) FROM SalePayment 
                              WHERE SalePayment.salePaymentSaleUid = sl.saleUid 
                               AND CAST(SalePayment.salePaymentDone AS INTEGER) = 1 
                               AND CAST(SalePayment.salePaymentActive AS INTEGER) = 1 ) ,
                          0)
                    ) > 0
                    ELSE CAST(sl.saleActive AS INTEGER) = 1 END
                
                GROUP BY saleUid, Customer.personUid 
                
                ORDER BY saleLastUpdateDate DESC
        """

    }
}
